python/sdist/amici/sbml_import.py

```diff
@@ -2785,58 +2801,156 @@ def _clean_reserved_symbols(self) -> None:
-        math_string = replace_logx(math_string)
-        try:
+
+        # already a sympy object
+        if isinstance(var_or_math, sp.Basic):
+            # substitute free symbols to match assumptions of other model
+            #  entities where necessary
+            ele_name = "SymPy expression"
+            expr = var_or_math.subs(
+                {
+                    sym: local
+                    for sym in var_or_math.free_symbols
+                    if (local := self._local_symbols.get(str(sym), sym)) != sym
+                }
+            )
+
+        # an expression string
+        elif isinstance(var_or_math, str):
+            if not var_or_math:
+                return None
+
+            ele_name = "Expression string"
             try:
-                formula = sp.sympify(
-                    _parse_logical_operators(math_string),
-                    locals=self._local_symbols,
+                expr = sp.sympify(
+                    var_or_math,
+                    locals={
+                        "true": sp.Float(1.0),
+                        "false": sp.Float(0.0),
+                        **self._local_symbols,
+                    },
                 )
-            except TypeError as err:
-                if str(err) == "BooleanAtom not allowed in this context.":
-                    formula = sp.sympify(
-                        _parse_logical_operators(math_string),
-                        locals={
-                            "true": sp.Float(1.0),
-                            "false": sp.Float(0.0),
-                            **self._local_symbols,
-                        },
-                    )
-                else:
-                    raise
-        except (sp.SympifyError, TypeError, ZeroDivisionError) as err:
-            raise SBMLException(
-                f'{ele_name} "{math_string}" '
-                "contains an unsupported expression: "
-                f"{err}."
+            except (sp.SympifyError, TypeError, ZeroDivisionError) as err:
+                raise SBMLException(
+                    f'{ele_name} "{var_or_math}" '
+                    "contains an unsupported expression: "
+                    f"{err}."
+                )
+
+        # actual libSBML objects
+        elif isinstance(var_or_math, libsbml.ASTNode | libsbml.SBase):
+            if isinstance(var_or_math, libsbml.ASTNode):
+                ast_node = var_or_math
+                sbml_obj = ast_node.getParentSBMLObject()
+            elif isinstance(var_or_math, libsbml.SBase):
+                sbml_obj = var_or_math
+                ast_node = var_or_math.getMath()
+            ele_name = sbml_obj.getId()
+
+            mathml = libsbml.writeMathMLWithNamespaceToString(
+                ast_node,
+                libsbml.SBMLNamespaces(
+                    self.sbml.getLevel(), self.sbml.getVersion()
+                ),
             )
+            try:
+                expr = self._mathml_parser.parse_str(mathml)
```

Comment:
same as above, make this one big try clause?

Reply:
I don't think `.subs()` will raise, I'd leave it out of there.